### PR TITLE
[JBEAP-4036] "number-of-application-rollbacks" statistic incremented twice

### PR DESCRIPTION
**Pull request by TomasHofman**
…during single rollback

https://issues.jboss.org/browse/JBTM-2643
https://issues.jboss.org/browse/JBEAP-4036

!BLACKTIE !XTS !QA_JTA !QA_JTS_JACORB !QA_JTS_JDKORB !QA_JTS_IBMORB

### CODE REVIEW

File: ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TwoPhaseCoordinator.java [hunk 1]
```diff
@@ -116,17 +116,20 @@ public int end (boolean report_heuristics)
...
-			TxStats.getInstance().incrementApplicationRollbacks();
-
 		if (parent() != null)
 			parent().removeChildAction(this);
 
 		// beforeCompletion();
 
-		int outcome = super.Abort();
+		int outcome;
+		synchronized (_cancelLock) { // https://issues.jboss.org/browse/JBTM-2643
```
Comment (mmusgrov) [inline ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TwoPhaseCoordinator.java]: I don't like having yet another lock in here. Why not put the conditional stat increment in BasicAction::Abort which is already synchronized.

Comment (TomasHofman) [inline ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TwoPhaseCoordinator.java]: OK, sounds better. I will update the PR.

Comment (mmusgrov) [inline ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TwoPhaseCoordinator.java]: Something like:
https://github.com/mmusgrov/narayana/commit/39669f18f5c2a06aacac030b018416b010d8aea0

**Comment by TomasHofman** [comment]
Updated with Michael's solution.

**Comment by TomasHofman** [comment]
@tomjenkinson I signed. Is that necessary even for Red Hat employees, and does it apply for all projects, or is it a JBoss TM specific thing? Thanks!

**Comment by mmusgrov** [comment]
The changes look good, thanks for raising this issue. Before you can commit the changes you need to rebase your tree.

**Comment by TomasHofman** [comment]
Then I need to open a new PR against master - this one I opened against 5.2 branch and it's up-to-date already.

**Comment by mmusgrov** [comment]
The fix first needs to be on master and then if it needs to be back ported then that would be a separate discussion.

**Comment by TomasHofman** [comment]
Opened https://github.com/jbosstm/narayana/pull/993, I'm closing this PR then.

**Comment by TomasHofman** [comment]
Reopening this in order to backport to 5.2 branch.

**Pull request closed without merging**